### PR TITLE
Make NuidWriter public

### LINKED ISSUE #617: Make the NuidWriter class public
### Proposed change

Move `NuidWriter` from `NATS.Client.Core.Internal` to `NATS.Client.Core` and make the class `public` instead of `internal`.

### Use case

I use Nuids for identifiers throughout my application. The Nuid generator is public in v1 of the client, so it makes sense for it to be so in v2.

### Contribution

I'm creating this issue for tracking purposes from the [original discussion post](https://github.com/nats-io/nats.net/discussions/616), PR incoming shortly.

### PR DESCRIPTION
**Pull request by Madgvox**
This PR makes the `NuidWriter` class public, and moves it into the `Core` namespace out of the `Internal` namespace. 

Resolves #617.

### CODE REVIEW

**Comment by mtmk** [comment]
Thanks @Madgvox. Can I ask you to sign your commit please? It's a CNCF requirement, and we won't be able to merge because of GitHub rules setup for this repo. You will have to force push, which should be fine. Thank you 💯

**Comment by Madgvox** [comment]
> Thanks @Madgvox. Can I ask you to sign your commit please? It's a CNCF requirement, and we won't be able to merge because of GitHub rules setup for this repo. You will have to force push, which should be fine. Thank you 💯

Should be corrected now.

**Comment by jasper-d** [comment]
I think it would be better change the naming as follows:

```diff
- public static class NuidWriter {
+ public static class Nuid { // or NuidGenerator if you prefer
-   public static bool TryWriteNuid(Span<char> nuidBuffer);
+   public static int Write(Span<char> destination); // Throw for inappropriate sized destination, return number of chars written
}
```

Reasoning:

1. Having `GetString()` on a type that is call "writer" seems a bit odd, calling the type `Nuid` or `NuidGenerator` avoids that issue.
2. `TryWriteNuid` does not provide information how many chars are written, this is internal knowledge of the implementation.
3. `NuidWriter.TryWriteNuid` duplicates notions of "nuid" and "write" unnecessarily
4. I would expect callers to usually know how large their buffer is, so the try-pattern seems a bit unecessary here.
5. If you want to keep the try-pattern (may possibly provide some perf-benefits with enough inlining and a suitable call-site), I'd propose `bool TryWrite(Span<char> destination, out int charsWritten)` to align better with other try-APIs in the BCL.
6. Additionally, you may want to expose `NuidLength` as `public static readonly int Length` or something, so callers can use it when allocating buffers instead of relying on undocumented implementation details.

**Comment by mtmk** [comment]
> I think it would be better change the naming as follows:
> 
> ```diff
> - public static class NuidWriter {
> + public static class Nuid { // or NuidGenerator if you prefer
> -   public static bool TryWriteNuid(Span<char> nuidBuffer);
> +   public static int Write(Span<char> destination); // Throw for inappropriate sized destination, return number of chars written
> }
> ```
> 
> Reasoning:
> 
> 1. Having `GetString()` on a type that is call "writer" seems a bit odd, calling the type `Nuid` or `NuidGenerator` avoids that issue.

I'd be happy with either name

> 2. `TryWriteNuid` does not provide information how many chars are written, this is internal knowledge of the implementation.
> 3. `NuidWriter.TryWriteNuid` duplicates notions of "nuid" and "write" unnecessarily
> 4. I would expect callers to usually know how large their buffer is, so the try-pattern seems a bit unecessary here.
> 5. If you want to keep the try-pattern (may possibly provide some perf-benefits with enough inlining and a suitable call-site), I'd propose `bool TryWrite(Span<char> destination, out int charsWritten)` to align better with other try-APIs in the BCL.

This makes a lot of sense if we want to expose it. @Madgvox would you be using `TryWrite`? 

> 6. Additionally, you may want to expose `NuidLength` as `public static readonly int Length` or something, so callers can use it when allocating buffers instead of relying on undocumented implementation details.

If we want `TryWrite` it'd be good to expose `Length`

**Comment by Madgvox** [comment]
I'd be using the GetString() variant most likely. We use Nuids for dynamic
subject IDs.

**Comment by mtmk** [comment]
Renamed the class to `Nuid` and made TryWriteNuid internal again. we can open it up if there is need. plus xml docs. @Madgvox @jasper-d let me know if you're happy then I shall merge. thank you 💯

File: src/NATS.Client.Core/Nuid.cs [hunk 1]
```diff
@@ -42,25 +49,30 @@ private NuidWriter()
...
-    public static bool TryWriteNuid(Span<char> nuidBuffer)
+    /// <summary>
+    /// Generates a new NATS unique identifier (NUID).
+    /// </summary>
+    /// <returns>A new NUID as a string.</returns>
+    /// <exception cref="InvalidOperationException">Thrown when unable to generate the NUID.</exception>
+    public static string NewNuid()
     {
-        if (_writer is not null)
+        Span<char> buffer = stackalloc char[22];
```
Comment (jasper-d) [inline src/NATS.Client.Core/Nuid.cs:59]: nit:

```suggestion
        Span<char> buffer = stackalloc char[NuidLength];
```

**Review by jasper-d — APPROVED** [review]
Code changes LGTM, did not review workflow changes.

**Comment by Madgvox** [comment]
LGTM 👍

**Review by mtmk — APPROVED** [review]
LGTM thanks @Madgvox and @jasper-d

**Pull request merged**